python/src/trezorlib/stellar.py

```diff
@@ -33,309 +55,215 @@
-
-        tx.timebounds_start = unpacker.unpack_uhyper()
-        tx.timebounds_end = unpacker.unpack_uhyper()
-
-        if tx.timebounds_start > max_timebound or tx.timebounds_start < 0:
-            raise ValueError(
-                "Starting timebound out of range (must be between 0 and "
-                + max_timebound
-            )
-        if tx.timebounds_end > max_timebound or tx.timebounds_end < 0:
-            raise ValueError(
-                "Ending timebound out of range (must be between 0 and " + max_timebound
-            )
-
-    # memo type determines what optional fields are set
-    tx.memo_type = unpacker.unpack_uint()
-
-    # text
-    if tx.memo_type == MEMO_TYPE_TEXT:
-        tx.memo_text = unpacker.unpack_string().decode()
-    # id (64-bit uint)
-    if tx.memo_type == MEMO_TYPE_ID:
-        tx.memo_id = unpacker.unpack_uhyper()
-    # hash / return are the same structure (32 bytes representing a hash)
-    if tx.memo_type == MEMO_TYPE_HASH or tx.memo_type == MEMO_TYPE_RETURN:
-        tx.memo_hash = unpacker.unpack_fopaque(32)
-
-    tx.num_operations = unpacker.unpack_uint()
+    parsed_tx = parsed_te.transaction
+    tx.source_account = parsed_tx.source.account_id
+    tx.fee = parsed_tx.fee
+    tx.sequence_number = parsed_tx.sequence
 
+    # Timebounds is an optional field
+    if parsed_tx.time_bounds:
+        tx.timebounds_start = parsed_tx.time_bounds.min_time
+        tx.timebounds_end = parsed_tx.time_bounds.max_time
+
+    memo = parsed_tx.memo
+    if isinstance(memo, TextMemo):
+        # memo_text is specified as UTF-8 string, but returned as bytes from the XDR parser
+        tx.memo_type = MEMO_TYPE_TEXT
+        tx.memo_text = memo.memo_text.decode("utf-8")
+    elif isinstance(memo, IdMemo):
+        tx.memo_type = MEMO_TYPE_ID
+        tx.memo_id = memo.memo_id
+    elif isinstance(memo, HashMemo):
+        tx.memo_type = MEMO_TYPE_HASH
+        tx.memo_hash = memo.memo_hash
+    elif isinstance(memo, ReturnHashMemo):
+        tx.memo_type = MEMO_TYPE_RETURN
+        tx.memo_hash = memo.memo_return
+    else:
+        tx.memo_type = MEMO_TYPE_NONE
+
+    tx.num_operations = len(parsed_tx.operations)
     operations = []
-    for _ in range(tx.num_operations):
-        operations.append(_parse_operation_bytes(unpacker))
+    for parsed_operation in parsed_tx.operations:
+        operations.append(_read_operation(parsed_operation))
 
     return tx, operations
 
 
-def _parse_operation_bytes(unpacker):
-    """Returns a protobuf message representing the next operation as read from
-    the byte stream in unpacker
-    """
-
-    # Check for and parse optional source account field
-    source_account = None
-    if unpacker.unpack_bool():
-        source_account = unpacker.unpack_fopaque(32)
-
-    # Operation type (See OP_ constants)
-    type = unpacker.unpack_uint()
-
-    if type == OP_CREATE_ACCOUNT:
+def _read_operation(parsed_operation: Operation):
```

Comment:
suggest renaming `parsed_operation` to `op`, it's significantly shorter and still very clear

Reply:
75ebbce0d081dde71c79022c1822943c892efe32